s3_storage_provider.py

```diff
@@ -127,34 +115,29 @@ def _get_s3_client(self):
     def store_file(self, path, file_info):
         """See StorageProvider.store_file"""
 
-        parent_logcontext = current_context()
-
         def _store_file():
-            with LoggingContext(parent_context=parent_logcontext):
-                self._get_s3_client().upload_file(
-                    Filename=os.path.join(self.cache_directory, path),
-                    Bucket=self.bucket,
-                    Key=self.prefix + path,
-                    ExtraArgs=self.extra_args,
-                )
-
-        return make_deferred_yieldable(
-            threads.deferToThreadPool(reactor, self._s3_pool, _store_file)
+            self._get_s3_client().upload_file(
+                Filename=os.path.join(self.cache_directory, path),
+                Bucket=self.bucket,
+                Key=self.prefix + path,
+                ExtraArgs=self.extra_args,
+            )
+
+        return run_in_background(
+            threads.deferToThreadPool, reactor, self._s3_pool, _store_file
```

Comment:
While `run_in_background` will run `threads.deferToThreadPool` in the same logcontext, `threads.deferToThreadPool` will not run `_store_file` in any sort of logcontext.

What we really need is `synapse.logging.context.defer_to_threadpool(...)` which basically did the same pattern as before.

It looks like we do export this pattern via `ModuleApi.defer_to_thread(...)` which means we can also get rid of the `self._s3_pool`

```suggestion
        return module_api.defer_to_thread(reactor, self._s3_pool, _store_file)
```

Reply:
I've just realised we don't actually pass an instance of `ModuleApi` to `media_storage_providers`...

We could do so [here](https://github.com/element-hq/synapse/blob/50b74bac932def1f6970c82da3c09d38105b9a19/synapse/media/media_repository.py#L135), and honestly I see little reason not to. Not doing so hobbles what media storage providers can do.

Reply:
We could use `hs.get_module_api()`

Reply:
Actually, we can get it from `hs.get_module_api()`. Though I wonder whether we should take this as an opportunity to update the signature of `__init__` completely to only be `__init(config, module_api)`...

Reply:
@anoadragon453 Haven't put enough thought into it.

If we go with `hs.get_module_api()` to move forwards now, it doesn't stop us from making a holistic change in the future.

Reply:
I've decided to go with `hs.get_module_api()` as there's already enough complication with this change.